kafka-native/src/main/java/org/ballerinalang/messaging/kafka/observability/KafkaTracingUtil.java

```diff
@@ -18,49 +18,41 @@
 
 package org.ballerinalang.messaging.kafka.observability;
 
+import io.ballerina.runtime.api.Environment;
 import io.ballerina.runtime.api.values.BObject;
 import io.ballerina.runtime.observability.ObserveUtils;
 import io.ballerina.runtime.observability.ObserverContext;
-import io.ballerina.runtime.scheduling.Strand;
 import org.ballerinalang.messaging.kafka.utils.KafkaUtils;
 
-import java.util.Optional;
-
 /**
  * Providing tracing functionality to Kafka.
  *
  * @since 1.2.0
  */
 public class KafkaTracingUtil {
 
-    public static void traceResourceInvocation(Strand strand, BObject object, String topic) {
+    public static void traceResourceInvocation(Environment environment, BObject object, String topic) {
         if (!ObserveUtils.isTracingEnabled()) {
             return;
         }
-        ObserverContext observerContext;
-        Optional<ObserverContext> observerContextOptional = ObserveUtils.getObserverContextOfCurrentFrame(strand);
-        if (observerContextOptional.isPresent()) {
-            observerContext = observerContextOptional.get();
-        } else {
+        ObserverContext observerContext = ObserveUtils.getObserverContextOfCurrentFrame(environment);
+        if (observerContext == null) {
             observerContext = new ObserverContext();
-            ObserveUtils.setObserverContextToCurrentFrame(strand, observerContext);
+            ObserveUtils.setObserverContextToCurrentFrame(environment, observerContext);
         }
         setTags(observerContext, object, topic);
     }
 
-    public static void traceResourceInvocation(Strand strand, BObject object) {
+    public static void traceResourceInvocation(Environment environment, BObject object) {
         if (!ObserveUtils.isTracingEnabled()) {
             return;
         }
-        ObserverContext observerContext;
-        Optional<ObserverContext> observerContextOptional = ObserveUtils.getObserverContextOfCurrentFrame(strand);
-        if (observerContextOptional.isPresent()) {
-            observerContext = observerContextOptional.get();
-        } else {
+        ObserverContext observerContext = ObserveUtils.getObserverContextOfCurrentFrame(environment);
+        if (observerContext == null) {
             observerContext = new ObserverContext();
-            ObserveUtils.setObserverContextToCurrentFrame(strand, observerContext);
+            ObserveUtils.setObserverContextToCurrentFrame(environment, observerContext);
+            setTags(observerContext, object);
```

Comment:
Still wrong. Move this out of the if clause. otherwise, you don't set tags when observerContext is not null